stock_dynamic_routing/models/stock_routing.py

```diff
@@ -0,0 +1,285 @@
+# Copyright 2019-2020 Camptocamp (https://www.camptocamp.com)
+# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+
+from collections import defaultdict
+from functools import lru_cache
+
+from odoo import _, api, fields, models
+from odoo.osv import expression
+
+
+def _default_sequence(model):
+    maxrule = model.search([], order="sequence desc", limit=1)
+    if maxrule:
+        return maxrule.sequence + 10
+    else:
+        return 0
+
+
+class StockRouting(models.Model):
+    _name = "stock.routing"
+    _description = "Stock Dynamic Routing"
+    _order = "sequence, id"
+
+    _rec_name = "location_id"
+
+    location_id = fields.Many2one(
+        comodel_name="stock.location",
+        required=True,
+        ondelete="cascade",
+        index=True,
+    )
+    picking_type_id = fields.Many2one(
+        comodel_name="stock.picking.type",
+        string="Operation Type",
+        required=True,
+        help="Apply this routing only if the operation type of the move is the same.",
+    )
+    sequence = fields.Integer(default=lambda self: self._default_sequence())
+    active = fields.Boolean(default=True)
+    rule_ids = fields.One2many(
+        string="Routing Rules",
+        comodel_name="stock.routing.rule",
+        inverse_name="routing_id",
+    )
+    routing_message = fields.Html(compute="_compute_routing_message")
+
+    _sql_constraints = [
+        (
+            "location_picking_type_uniq",
+            "unique(location_id, picking_type_id)",
+            "A routing configuration already exists for this location and picking type",
+        )
+    ]
+
+    def _routing_message_template(self):
+        pull_rules = self.rule_ids.filtered(lambda r: r.method == "pull")
+        push_rules = self.rule_ids.filtered(lambda r: r.method == "push")
+        pull_message = ""
+        push_message = ""
+
+        if pull_rules:
+            rule_messages = []
+            for rule in pull_rules:
+                msg = _(
```

Comment:
```suggestion
                msg = self.env._(
```
same for all gettext calls